src/net_processing.cpp

```diff
@@ -816,6 +816,91 @@ void Misbehaving(NodeId pnode, int howmuch, const std::string& message) EXCLUSIV
         LogPrint(BCLog::NET, "%s: %s peer=%d (%d -> %d)%s\n", __func__, state->name, pnode, state->nMisbehavior-howmuch, state->nMisbehavior, message_prefixed);
 }
 
+/**
+ * Returns true if the given validation state result may result in us banning/disconnecting a peer
+ * which provided such an object. This is used to determine whether to relay transactions to
+ * whitelisted peers, preventing us from relaying things which would result in them disconnecting
+ * us.
+ */
+static bool MayResultInDisconnect(const CValidationState& state, bool via_compact_block) {
+    switch (state.GetReason()) {
+    case ValidationInvalidReason::NONE:
+        return false;
+    // The node is providing invalid data:
+    case ValidationInvalidReason::CONSENSUS:
+    case ValidationInvalidReason::BLOCK_MUTATED:
+    case ValidationInvalidReason::CACHED_INVALID:
+        // Blocks (or headers) invalid for one of the above reasons may not be
+        // knowable to a high-bandwidth compact block peer, prior to relay.
+        // Headers that are invalid for reasons that should be known prior to
+        // block validation -- such as bad proof of work, too-early-time, or
+        // building off an invalid or missing block -- are punished regardless
+        // (see below).
+        return !via_compact_block;
+    case ValidationInvalidReason::BLOCK_INVALID_HEADER:
+    case ValidationInvalidReason::BLOCK_CHECKPOINT:
+    case ValidationInvalidReason::BLOCK_INVALID_PREV:
```

Comment:
It's unclear to me whether peers should always be punished for `BLOCK_INVALID_PREV` should be here. For example, if the previous block was invalid because of `RECENT_CONSENSUS_CHANGE` and the peer wasn't punished, should it be punished for relaying this descendant block?

Should compact block peers be punished for relaying the block if it's parent is invalid? My reading of https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki#pre-validation-relay-and-consistency-considerations is that the answer is no.

Same question for `MaybePunishNode()` below.

Reply:
If miners have mostly upgraded then building on top of a `RECENT_CONSENSUS_CHANGE` block should be rare enough for this not to be a huge problem.

If not, and we want to cope with a moderately controversial consensus upgrade, then we probably want to track whether blocks failed due to `RECENT_CONSENSUS_CHANGE` and mark their children as also failing due to `RECENT_CONSENSUS_CHANGE` (after checking PoW at least). Working all that out doesn't seem necessary for this patchset to me though.

Reply:
With respect to the issue you're bringing up, I believe the behavior in this PR matches existing behavior, in which case I'd prefer to defer improvement to a future PR.  If I'm missing some way that we've made things different or worse though let me know.

As for BIP 152:

>A node MUST NOT send a cmpctblock message without having validated that the header properly commits to each transaction in the block, and properly builds on top of the existing, fully-validated chain with a valid proof-of-work either as a part of the current most-work valid chain, or building directly on top of it. A node MAY send a cmpctblock before validating that each transaction in the block validly spends existing UTXO set entries.